File: OnlineTaksiSwift/OnlineTaksiSwift/util/SocketEventParser.swift
//
//  SocketEventParser.swift
//  OnlineTaksiSwift
//
//  Created by Alperen Saraç on 24.04.2026.
//

import Foundation

enum SocketEventParser {

    static func getEventName(_ message: String) -> String? {
        guard let data = message.data(using: .utf8),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            return nil
        }

        return json["event"] as? String
    }

    static func parseRideStatus(_ message: String) -> String? {
        guard let data = message.data(using: .utf8),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let event = json["event"] as? String,
              ["RIDE_ACCEPTED", "RIDE_STATUS_CHANGED", "RIDE_CANCELLED"].contains(event),
              let body = json["data"] as? [String: Any],
              let status = body["status"] as? String else {
            return nil
        }

        return status
    }

    static func parseDriverLocation(_ message: String) -> DriverLocationEvent? {
        guard let data = message.data(using: .utf8),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let event = json["event"] as? String,
              event == "DRIVER_LOCATION",
              let body = json["data"] as? [String: Any],
              let rideId = body["ride_id"] as? Int,
              let driverId = body["driver_id"] as? Int,
              let lat = body["lat"] as? Double,
              let lng = body["lng"] as? Double else {
            return nil
        }

        return DriverLocationEvent(
            rideId: rideId,
            driverId: driverId,
            lat: lat,
            lng: lng
        )
    }
}

File: OnlineTaksiKotlin/app/src/main/java/com/example/onlinetaksi/data/remote/socket/SocketEventParser.kt
package com.example.onlinetaksi.data.remote.socket

import com.example.onlinetaksi.data.remote.model.DriverLocationEvent
import org.json.JSONObject

object SocketEventParser {

    fun getEventName(message: String): String? {
        return try {
            val jsonObject = JSONObject(message)
            jsonObject.optString("event")
        } catch (e: Exception) {
            null
        }
    }

    fun parseDriverLocation(message: String): DriverLocationEvent? {
        return try {
            val jsonObject = JSONObject(message)
            val eventName = jsonObject.optString("event")
            if (eventName != "DRIVER_LOCATION") return null

            val data = jsonObject.optJSONObject("data") ?: return null

            DriverLocationEvent(
                rideId = data.optInt("ride_id", -1),
                driverId = data.optInt("driver_id", -1),
                lat = data.optDouble("lat", 0.0),
                lng = data.optDouble("lng", 0.0)
            )
        } catch (e: Exception) {
            null
        }
    }

    fun parseRideStatus(message: String): String? {
        return try {
            val jsonObject = JSONObject(message)
            val eventName = jsonObject.optString("event")

            if (eventName != "RIDE_STATUS_CHANGED" &&
                eventName != "RIDE_ACCEPTED" &&
                eventName != "RIDE_CANCELLED"
            ) {
                return null
            }

            val data = jsonObject.optJSONObject("data") ?: return null
            data.optString("status")
        } catch (e: Exception) {
            null
        }
    }
}